### PR TITLE
Firebase

### PR DESCRIPTION
**Pull request by arrtemio**

### CODE REVIEW

File: src/entities/Task/model/actions/tasksActions.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+} from '../types/task';
+
+const tasksCollection = collection(firestore, 'tasks');
+
+export const getTaskByID = async (taskID: string) => {
+    const taskRef = doc(tasksCollection, taskID);
+    const taskSnapshot = await getDoc(taskRef);
+    const taskData = taskSnapshot.data();
+
+    if (!taskData) throw new Error('task didnt find');
```
Comment (sshkatulov) [inline src/entities/Task/model/actions/tasksActions.ts]: 1. `task not found`
2. what's exacly supposed to happen when we throw an error here?

File: src/entities/Task/model/types/task.ts [hunk 2]
```diff
@@ -13,11 +15,43 @@ export interface BaseTask {
...
+export interface Task extends BaseTask {
+    createdAt: number;
+    subTasks: BaseTask[];
+}
+
+export interface BaseTaskDTO {
+    title: string;
+    completedAt: number | null;
+    status: Status;
+}
```
Comment (sshkatulov) [inline src/entities/Task/model/types/task.ts]: you just specified another BaseTask but without an `id`. imagine if there are like several dosens of props, too much copy-paste will be required. you could do something like this `export type BaseTaskDTO = Omit<BaseTask, 'id'>;`, however it looks like the better solution would be to extend BaseTask from BaseTaskDTO

File: src/entities/Task/model/actions/tasksActions.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    subTaskID: string;
+}
+
+export const changeSubTaskStatus = createAsyncThunk(
+    'tasks/changeSubTaskStatus',
+    async ({ status, taskID, subTaskID }: ChangeSubTaskStatus) => {
+        const { taskData, taskRef } = await getTaskByID(taskID);
+        const currentSubTask = taskData.subTasks.find((sub: BaseTask) => sub.id === subTaskID);
+
+        changeSubTaskStatusFn(taskData as Task, currentSubTask, status);
```
Comment (sshkatulov) [inline src/entities/Task/model/actions/tasksActions.ts]: `taskData as Task` happens three times in the file, maybe we can specify the returned type for `getTaskByID` istead??

File: src/entities/Task/model/slice/taskSlice.test.ts [hunk 4]
```diff
@@ -2,76 +2,94 @@ import { DeepPartial } from '@reduxjs/toolkit';
...
 
 describe('tasksSlice', () => {
-    const state: DeepPartial<TasksSchema> = {
-        data: [
-            testTask,
-        ],
-    };
+    let state: DeepPartial<TasksSchema>;
+    beforeEach(() => {
+        state = {
```
Comment (sshkatulov) [inline src/entities/Task/model/slice/taskSlice.test.ts]: in some tests you define new `const state`, so you don't actually need such `beforeEach` for them and such tests can be moved to another describe. also, for the rest of the tests you can get rid of `DeepPartial<>` in `DeepPartial<TasksSchema>` and `as TasksSchema` in `state as TasksSchema`

File: src/entities/Task/model/selectors/tasksSelectors.test.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import { DeepPartial } from '@reduxjs/toolkit';
+import { testTask } from 'shared/test/TestTask';
+import { StateSchema } from 'app/store';
+import { getTasksData, getTasksError, getTasksLoading } from './tasksSelectors';
+
+describe('tasksSelectors test', () => {
+    let state: DeepPartial<StateSchema>;
```
Comment (sshkatulov) [inline src/entities/Task/model/selectors/tasksSelectors.test.ts]: why not just 
```
    const state: StateSchema = {
        tasks: {
            data: [testTask],
            isLoading: true,
            error: 'Error',
        },
    };
```
for the first three tests intead of having `DeepPartial`?

File: src/entities/Task/model/types/task.ts [hunk 6]
```diff
@@ -13,11 +15,43 @@ export interface BaseTask {
...
+export const createSubTaskDto = (title: string): BaseTask => (
+    {
+        title,
+        completedAt: null,
+        status: Status.TO_DO,
+        id: generateRandomId(),
+    }
+);
+
+export const createTasKDto = (title: string): TaskDTO => (
```
Comment (sshkatulov) [inline src/entities/Task/model/types/task.ts]: `createTaskDto`

File: .env
Comment (sshkatulov) [inline .env:1]: even if it's an educational project let's remove the config from git

File: src/entities/Task/model/actions/tasksActions.ts (hunk 1 above)
Comment (arrtemio) [reply src/entities/Task/model/actions/tasksActions.ts]: Fixed

File: src/entities/Task/model/types/task.ts (hunk 2 above)
Comment (arrtemio) [reply src/entities/Task/model/types/task.ts]: Fixed

File: src/entities/Task/model/actions/tasksActions.ts (hunk 3 above)
Comment (arrtemio) [reply src/entities/Task/model/actions/tasksActions.ts]: Fixed

File: src/entities/Task/model/slice/taskSlice.test.ts (hunk 4 above)
Comment (arrtemio) [reply src/entities/Task/model/slice/taskSlice.test.ts]: Done

File: src/entities/Task/model/selectors/tasksSelectors.test.ts (hunk 5 above)
Comment (arrtemio) [reply src/entities/Task/model/selectors/tasksSelectors.test.ts]: Fixed

File: src/entities/Task/model/types/task.ts (hunk 6 above)
Comment (arrtemio) [reply src/entities/Task/model/types/task.ts]: Fixed

File: .env
Comment (arrtemio) [reply .env:1]: Removed

File: src/entities/Task/model/actions/tasksActions.ts (hunk 1 above)
Comment (arrtemio) [reply src/entities/Task/model/actions/tasksActions.ts]: This was supposed to terminate the execution of the function, for example, changeTaskStatus. However, as it turned out, it was simply breaking the application, so now rejectWithValue is used, which can catch this error without blocking the application.

**Review by sshkatulov — APPROVED** [review]

**Pull request merged**